src/state_db.rs

```diff
@@ -454,9 +451,293 @@ impl StateDB {
+                        } else {
+                            // if there is no key then make a new pair and insert into cache
+                            writer.cache_new(&SharedKVPair::new(&key, &new_value));
+                            vec![ctx.null().upcast(), JsNumber::new(&mut ctx, 3).upcast()]
+                        }
+                    } else {
+                        let err = value.err().unwrap();
+                        vec![ctx.error(&err)?.upcast()]
+                    }
+                };
+
+                let this = ctx.undefined();
+                let callback = cb.into_inner(&mut ctx);
+                callback.call(&mut ctx, this, args)?;
+                Ok(())
+            });
+        })
+    }
+
+    fn delete_key(
+        &self,
+        cb: Root<JsFunction>,
+        writer: ArcMutex<state_writer::StateWriter>,
+        key: Vec<u8>,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        let state_db_key = DBKind::State.key(key.clone());
+        self.common.send(move |conn, channel| {
+            let value = conn.get(&state_db_key);
+            channel.send(move |mut ctx| {
+                let this = ctx.undefined();
+                let callback = cb.into_inner(&mut ctx);
+                // the following scope use to release writer at the end of it
+                {
+                    let mut writer = writer.lock().unwrap();
+                    let cached = writer.is_cached(&key);
+                    if !cached {
+                        if let Ok(value) = &value {
+                            // if found the value of the key then insert into cache
+                            if value.is_some() {
+                                let temp_value = value.as_ref().unwrap().to_vec();
+                                let pair = SharedKVPair::new(&key, &temp_value);
+                                writer.cache_existing(&pair);
+                            }
+                        } else {
+                            let err = value.err().unwrap();
+                            let args = vec![ctx.error(&err)?.upcast()];
+                            callback.call(&mut ctx, this, args)?;
+                        }
+                    }
+                    writer.delete(&key);
+                }
+                let args = vec![ctx.null().upcast()];
+                callback.call(&mut ctx, this, args)?;
+                Ok(())
+            });
+        })
+    }
+
+    fn get_key_with_writer(
+        &self,
+        cb: Root<JsFunction>,
+        writer: ArcMutex<state_writer::StateWriter>,
+        key: Vec<u8>,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        let state_db_key = DBKind::State.key(key.clone());
+        self.common.send(move |conn, channel| {
+            let value = conn.get(&state_db_key);
+            channel.send(move |mut ctx| {
+                let args = {
+                    let mut writer = writer.lock().unwrap();
```

Comment:
You may release writer in similar way which is done in `delete_key` function.

Reply:
Same as above comment